scheduling-framework.md: update Reserve and Unreserve descriptions (#22038)
kubernetes/website: content/en/docs/concepts/scheduling-eviction/scheduling-framework.md

NormalizeScore extension point.
{{< /note >}}

### Reserve {#reserve}

The Reserve extension point happens before the scheduler actually binds a Pod to
a node. Its uses are three-fold. First, it serves as a hook to allocate
resources in the designated node that the scheduled Pod might need. Second, it
serves as an informational plugin which maintain runtime state (aka "stateful
plugins") to be notified by the scheduler when resources on a node are being
reserved for a given Pod. Third, it exists to prevent race conditions while the
scheduler waits for the bind to succeed. This is the penultimate step in the

> scheduling cycle.

Reviewer: s/penultimate/last

Author: see https://github.com/kubernetes/website/pull/22035#discussion_r444983940

and `Unreserve`, that back two informational scheduling phases called Reserve
and Unreserve, respectively. Plugins which maintain runtime state (aka "stateful
plugins") should use these phases to be notified by the scheduler when resources
on a node are being reserved and unreserved for a given Pod.

The Reserve phase happens before the scheduler actually binds a Pod to its
designated node. It exists to prevent race conditions while the scheduler waits
for the bind to succeed. The `Reserve` method of each Reserve plugin may succeed
or fail; if one `Reserve` method call fails, subsequent plugins are not executed
and the Reserve phase is considered to have failed. If the `Reserve` method of
all plugins succeed, the Reserve phase is considered to be successful and the
rest of the scheduling cycle and the binding cycle are executed.

The Unreserve phase is triggered if the Reserve phase or a later phase fails.
When this happens, the `Unreserve` method of **all** Reserve plugins will be
executed in the reverse order of `Reserve` method calls. This phase exists to
clean up the state associated with the reserved Pod. The implementation of the

> `Unreserve` method in Reserve plugins must be idempotent and may not fail.

Reviewer 2: >  The implementation of the `Unreserve` method in Reserve plugins must be idempotent and may not fail.

*Optional tweak*
This could be in a `{{< caution >}}` shortcode, or you could move that detail into a page inside https://k8s.io/docs/reference/scheduling/ and then signpost there with text like “if you're implementing a plugin, see xxx for more details and caveats”

Author: Good call. Updated with a caution tag.

NormalizeScore extension point.
{{< /note >}}

### Reserve {#reserve}

A Reserve plugin implements two methods, namely Reserve and Unreserve, to serve

> as informational extension points. Plugins which maintain runtime state (aka

Reviewer 3: let's say something like (@Huang-Wei for thoughts):

The Reserve extension point backs two informational scheduling phases: Reserve and Unreserve. Plugins implement two methods, Reserve and Unreserve, that get called during these phases, respectively.

Sorry for the back and forth, but I don't find an easy way to express the fact that Unreserve is not an extension point.

Reviewer: it is a callback

Reviewer 4: LGTM, slightly reworded:

```
The Reserve extension point backs two informational scheduling phases: Reserve and Unreserve,
which are implemented in pairs and get called during these phases, respectively.
```

Author: Done. I've slightly modified @Huang-Wei's suggestion to remove the respectively part as I don't think it's necessary (as there aren't two pairs of objects like in @alculquicondor's suggestion). PTAL.

Author: Also, we can explicitly state that Unreserve is not an extension point in addition to these changes to make it very clear?

> The Reserve extension point backs two informational scheduling phases called
> Reserve and Unreserve which are implemented in pairs and get called during these

Reviewer 3: ```suggestion
A plugin that implements the Reserve extension has two methods, namely `Reserve` and `Unreserve`
that back two informational scheduling phases, `Reserve` and `Unreserve`, respectively.
```

TBH, `implemented in pairs` sounds very confusing to me.

Author: Done.

phases. Plugins which maintain runtime state (aka "stateful plugins") should use

> these extension points to be notified by the scheduler when resources on a node

Author: Done.

are being reserved and unreserved for a given Pod.

The Reserve phase happens before the scheduler actually binds the Pod to the
designated node. It exists to prevent race conditions while the scheduler waits
for the bind to succeed. The Reserve method of each Reserve plugin may succeed
or fail; if one plugin fails, subsequent plugins are not executed and the

> extension point is considered to have failed. If all plugins' Reserve methods

Reviewer 3: ```suggestion
phase is considered to have failed. If all plugins' Reserve methods
```

Author: Done.

succeed, the extension point is considered to be successful and the rest of the
scheduling cycle and the binding cycle are executed.

> The Unreserve method of a Reserve plugin is executed if the Reserve extension

Reviewer 3: Something along the lines of `The Unreserve phase is triggered if the Reserve or a later phase fails`

Author: Done.

### Reserve {#reserve}

> A reserve plugin implements two methods, namely Reserve and Unreserve, to serve

Reviewer 3: ```suggestion
A Reserve plugin implements two methods, namely Reserve and Unreserve, to serve
```

Author: Done.

as informational extension points. Plugins which maintain runtime state (aka
"stateful plugins") should use these extension points to be notified by the
scheduler when resources on a node are being reserved and unreserved for a given
Pod.

> The Reserve extension point, which is the penultimate step of the scheduling
> cycle, is triggered before the scheduler actually binds the Pod to the

Reviewer 3: ```suggestion
The Reserve phase happens before the scheduler actually binds the Pod to the
```

I think there is no need to say penultimate, given that it can be deduced from the description of `Permit`.

Author: Done.

1.  **deny** \
    If any Permit plugin denies a Pod, it is returned to the scheduling queue.

>     This will trigger the [Unreserve](#unreserve) extension point.

Reviewer 4: Once the `Unreserve` stanza is moved, we can rephrase here:

```
This will trigger the "unreserve" API in the [Reserve](#reserve) extension point.
```

Reviewer 4: API can be replaced with other terms like "method" or "function".

Author: Done.

> ### Unreserve

Reviewer 4: Technically, `Unreserve` is not an extension point anymore. I'd suggest moving (and may need some rephrasing) this section as a bullet point of `Reserve`. So there can be two bullet points in `Reserve`, explaining the implementation/execution of `Reserve` and `Reserve` accordingly, under different circumstances.

Reviewer 3: Maybe we can call them "calls" of the "Reserve extension point"

Author: Unreserve isn't a plugin anymore, but isn't it still an extension point? It's just that one plugin triggers two extension points now. IMO, Reserve has no special meaning or difference compared to Unreserve; so they're both extension points or they're both not. Also, users can choose to leave the Reserve implementation empty. I think having a separate Unreserve section still makes sense, but I'd like to hear your thoughts.

Reviewer 4: I'd read the term "extension point" as an API defined in https://github.com/kubernetes/kubernetes/blob/bc60bdaded587ef638c165c49680728a57326f29/staging/src/k8s.io/kube-scheduler/config/v1beta1/types.go#L159, and also it should be 1:1 mapped to the rectangles in the architecture picture. Specifically, a plugin can implement multiple extension points, and an extension point represents a phase in a Pod's scheduling workflow. 

Most "extension points" just have one method defined, such as Fitler/Score. We do also have other ones such as PreFilter which has method "PreFilter" as well as "AddPod"/"RemovePod", implicitly though.

Reviewer 3: so, UnReserve is a scheduling phase, but not an extension point.

Reviewer 4: > UnReserve is a scheduling phase, but not an extension point.

I prefer to read it this way, and it's a _dynamic_ phase, that's why categorized it with a _fixed_ phase "Reserve" in pairs.

Author: >I'd read the term "extension point" as an API defined in https://github.com/kubernetes/kubernetes/blob/bc60bdaded587ef638c165c49680728a57326f29/staging/src/k8s.io/kube-scheduler/config/v1beta1/types.go#L159

Fair enough, that's a good point.

Done.

The Unreserve extension point is implemented by reserve plugins as a
complementary method to the Reserve extension point, and should clean up the
state associated with the reserved Pod. Triggering the unreserve extension point

> will cause the scheduler will execute the Unreserve method of **all** reserve

Reviewer: ```suggestion
will cause the scheduler to execute the Unreserve method of **all** reserve
```
